internal/build/imgsrc/ensure_builder.go

```diff
@@ -0,0 +1,325 @@
+		err := client.DeleteApp(ctx, builderApp.Name)
+		if err != nil {
+			tracing.RecordError(span, err, "error deleting invalid builder app")
+			return nil, nil, err
+		}
+	}
+
+	builderName := "fly-builder-" + haikunator.Haikunator().Build()
+	// we want to lauch the machine to the builder
+	flapsClient, err := flapsutil.NewClientWithOptions(ctx, flaps.NewClientOpts{
+		AppName: builderName,
+		OrgSlug: org.Slug,
+	})
+	if err != nil {
+		tracing.RecordError(span, err, "error creating flaps client")
+		return nil, nil, err
+	}
+	ctx = flapsutil.NewContextWithClient(ctx, flapsClient)
+	app, machine, err := createBuilder(ctx, org, region, builderName)
+	return machine, app, err
+}
+
+type ValidateBuilderError int
+
+func (e ValidateBuilderError) Error() string {
+	switch e {
+	case NoBuilderApp:
+		return "no builder app"
+	case NoBuilderVolume:
+		return "no builder volume"
+	case InvalidMachineCount:
+		return "invalid machine count"
+	default:
+		return "unknown error validating builder"
+	}
+}
+
+const (
+	NoBuilderApp ValidateBuilderError = iota
+	NoBuilderVolume
+	InvalidMachineCount
+)
+
+func validateBuilder(ctx context.Context, app *fly.App) (*fly.Machine, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "validate_builder")
+	defer span.End()
+
+	if app == nil {
+		tracing.RecordError(span, NoBuilderApp, "no builder app")
+		return nil, NoBuilderApp
+	}
+
+	flapsClient := flapsutil.ClientFromContext(ctx)
+
+	var volumes []fly.Volume
+	numRetries := 0
+
+	for {
+		var err error
+
+		volumes, err = flapsClient.GetVolumes(ctx)
+		if err == nil {
+			break
+		}
+
+		var flapsErr *flaps.FlapsError
+		// if it isn't a server error, no point in retrying
+		if errors.As(err, &flapsErr) && flapsErr.ResponseStatusCode >= 500 && flapsErr.ResponseStatusCode < 600 {
+			span.AddEvent(fmt.Sprintf("non-server error %d", flapsErr.ResponseStatusCode))
+			numRetries += 1
+
+			if numRetries >= 3 {
+				tracing.RecordError(span, err, "error getting volumes")
+				return nil, err
+			}
+			time.Sleep(1 * time.Second)
+		} else {
+			tracing.RecordError(span, err, "error getting volumes")
+			return nil, err
+		}
+	}
+
+	if len(volumes) == 0 {
+		tracing.RecordError(span, NoBuilderVolume, "the existing builder app has no volume")
+		return nil, NoBuilderVolume
+	}
+
+	var machines []*fly.Machine
+	for {
+		var err error
+
+		machines, err = flapsClient.List(ctx, "")
+		if err == nil {
+			break
+		}
+
+		var flapsErr *flaps.FlapsError
+		// if it isn't a server error, no point in retrying
+		if errors.As(err, &flapsErr) && flapsErr.ResponseStatusCode >= 500 && flapsErr.ResponseStatusCode < 600 {
+			span.AddEvent(fmt.Sprintf("non-server error %d", flapsErr.ResponseStatusCode))
+			numRetries += 1
+
+			if numRetries >= 3 {
+				tracing.RecordError(span, err, "error listing machines")
+				return nil, err
+			}
+			time.Sleep(1 * time.Second)
+		} else {
+			tracing.RecordError(span, err, "error listing machines")
+			return nil, err
+		}
+	}
```

Comment:
The volume handling and the machine handling could be split off into their own smaller functions as this function is getting kinda large.